packages/vector_graphics_compiler/lib/src/svg/overdraw_optimizer.dart

```diff
@@ -0,0 +1,310 @@
+          paint: bottomPathNode.paint,
+          bounds: bottomPathNode.bounds,
+          path: newBottomVGPath);
+      final ResolvedPathNode newTopPathNode = ResolvedPathNode(
+          paint: topPathNode.paint,
+          bounds: bottomPathNode.bounds,
+          path: newTopVGPath);
+      final ResolvedPathNode newOverlapPathNode = ResolvedPathNode(
+          paint: Paint(
+              blendMode: bottomPathNode.paint.blendMode,
+              stroke: bottomPathNode.paint.stroke,
+              fill: Fill(
+                  color: overlapColor,
+                  shader: bottomPathNode.paint.fill?.shader)),
+          bounds: bottomPathNode.bounds,
+          path: newOverlapVGPath);
+
+      bottomPathOpsPath.dispose();
+      topPathOpsPath.dispose();
+      intersection.dispose();
+      newBottomPath.dispose();
+
+      return <ResolvedPathNode>[
+        newBottomPathNode,
+        newTopPathNode,
+        newOverlapPathNode
+      ];
+    } else {
+      return <ResolvedPathNode>[bottomPathNode, topPathNode];
+    }
+  }
+
+  @override
+  _Result visitEmptyNode(Node node, void data) {
+    final _Result _result = _Result(node);
+    return _result;
+  }
+
+  /// Visits applies optimizer to all children of ParentNode.
+  _Result visitChildren(Node node, _Result data) {
+    if (node is ParentNode) {
+      data = node.accept(this, data);
+    }
+    return data;
+  }
+
+  @override
+  _Result visitParentNode(ParentNode parentNode, Node data) {
+    int pathNodeCount = 0;
+    final List<List<Node>> newChildList = <List<Node>>[];
+    List<Node> newChildren = <Node>[];
+
+    for (Node child in parentNode.children) {
+      if (child is ResolvedPathNode) {
+        pathNodeCount++;
+      }
+      newChildList.add(<Node>[child]);
+    }
+
+    int index = 0;
+    ResolvedPathNode? lastPathNode;
+    int? lastPathNodeIndex;
+    if (parentNode.attributes.opacity == null) {
+      if (pathNodeCount >= 2) {
+        for (Node child in parentNode.children) {
+          if (child is ResolvedPathNode &&
+              child.paint.stroke?.width == null &&
+              child.paint.stroke?.color == null &&
+              child.paint.fill?.shader == null) {
+            if (lastPathNode == null || lastPathNodeIndex == null) {
+              lastPathNode = child;
+              lastPathNodeIndex = index;
+            } else {
+              if (child.paint.fill?.color.a == 255) {
+                newChildList[lastPathNodeIndex] = <Node>[
+                  removeOverlap(lastPathNode, child)
+                ];
+              } else {
+                newChildList[lastPathNodeIndex] = resolveOpacityOverlap(
+                    (newChildList[lastPathNodeIndex].first as ResolvedPathNode),
+                    child);
+                newChildList[index] = <Node>[];
+                lastPathNode = null;
+                lastPathNodeIndex = null;
+              }
+            }
+          }
+          index++;
```

Comment:
This code is pretty difficult for me to follow, because there are a lot of `ifs` and several layers of nesting. Consider restructuring this a bit to make the logic clearer, either by splitting things off into private methods, adding early returns, or collapsing some of these checks into a single condition.

Also please add some comments describing the different paths this code can follow

Reply:
simplified logic and added some comments.